_posts/2020-08-05-using-open-policy-agent-with-strimzi-and-apache-kafka.md

```diff
@@ -0,0 +1,451 @@
+For each of these groups, we will need to define the set of operations they will be allowed to use.
+We will also add some helper rules to help us evaluate whether the operation we are authorizing belongs into one of these groups or not:
+
+```rego
+consumer_operations = {
+                        "Topic": ["Read", "Describe"], 
+                        "Group": ["Read", "Describe"]
+                    }
+
+producer_operations = {
+                        "Topic": ["Write", "Describe"]
+                    }
+
+admin_operations = {
+                    "Topic": ["Read", "Write", "Create", "Delete", "Alter", "Describe", "ClusterAction", "DescribeConfigs", "AlterConfigs"], 
+                    "Group": ["Read", "Delete", "Describe"],
+                    "Cluster": ["Create", "Alter", "Describe", "ClusterAction", "DescribeConfigs", "AlterConfigs"]
+                    }
+
+is_consumer_operation {
+    consumer_operations[input.resource.resourceType.name][_] == input.operation.name
+}
+
+is_producer_operation {
+    producer_operations[input.resource.resourceType.name][_] == input.operation.name
+}
+
+is_admin_operation {
+    admin_operations[input.resource.resourceType.name][_] == input.operation.name
+}
+```
+
+We also need to define the actual groups and their members:
+
+```rego
+consumer_group = ["tom", "tyrone", "matt", "pepe", "douglas"]
+producer_group = ["jack", "conor", "keinan", "john"]
+admin_group = ["dean", "christian"]
+
+is_consumer_group {
+    consumer_group[_] == principal.name
+}
+
+is_producer_group {
+    producer_group[_] == principal.name
+}
+
+is_admin_group {
+    admin_group[_] == principal.name
+}
+```
+
+The group definition is accompanied by some helper rules.
+Each group is defined as an array of its members.
+So we have the groups and their members hardcoded into the policy and to change them, you will need to change the policy.
+
+Next, we will create some helper rules for evaluating whether the request we are authorizing belongs to a user who is member of one of the groups and is for operation allowed for the group:
+
+```rego
+allow_consumer_group {
+    is_consumer_operation
+    is_consumer_group
+}
+
+allow_producer_group {
+    is_producer_operation
+    is_producer_group
+}
+
+allow_admin_group {
+    is_admin_operation
+    is_admin_group
+}
+```
+
+When the operation we are authorizing belongs to the consumer operations and the user belongs to the consumer group, the `allow_consumer_group` will evaluate to `true`.
+If the operation is not one of the operations allowed to consumers and/or the user doesn't belong to the consumer group, it will evaluate to `false`.
+The other two rules - `allow_producer_group` and `allow_admin_group` - work in the same way.
+
+Last, we need to add the `allow` rule which is what the Kafka Authorizer will actually call:
+
+```rego
+allow {
+    not deny
+}
+
+deny {
+    not allow_consumer_group
+    not allow_producer_group
+    not allow_admin_group
+}
+```
```

Comment:
as far as I understood in the OPA configuration you define the URL where you have the "rule" and in this case, it's `allow`. Why do we need to define the `deny` one, if it's not being used by the OPA authorizer?

Reply:
The `allow` is calling the `deny` rule. This is related to how the rules are evaluated in OPA and what Kafka expects as an answer from the authorizer => true (allowed) or false (denied) - so we have to basically invert the result of the `deny` rule.

Reply:
can't you just define `allow` allowing and not inverting `deny`? or OPA forces to have both `allow` and `deny`?

Reply:
Not to my knowledge (but it is not as if I'm Rego expert). 

To my understanding, when the rule consists from multiple sub-rules, they are handled as `rule AND rule AND ...`.  As a result, this:

```
deny {
    not allow_consumer_group
    not allow_producer_group
    not allow_admin_group
}
```

will evaluate as denied when the user is not allowed through any of the sub-rules. So when the user matches at least one of them, it will be allowed.

If you would do something like this instead:

```
allow {
   allow_consumer_group
   allow_producer_group
   allow_admin_group
}
```

it would allow the user only if all of these rules evaluate to true. Otherwise it would be denied. So the user would need to be valid consumer, producer and admin at the same time just to produce messages for example. So it would not work.

Reply:
ok it makes more sense now. Thanks.